File: 24.c
#include <stdio.h>
#include <stdlib.h>

// Definition for singly-linked list.
struct ListNode {
    int val;
    struct ListNode *next;
};

// Function to print the linked list
void printList(struct ListNode* head) {
    struct ListNode* curr = head;
    while (curr != NULL) {
        printf("%d", curr->val);
        if (curr->next != NULL)
            printf("->");
        curr = curr->next;
    }
    printf("\n");
}

struct ListNode* swapPairs(struct ListNode* head) {
    struct ListNode *temp = (struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode *prev = NULL;
    struct ListNode *curr = head;
    struct ListNode *next = NULL;
     
    if (head == NULL || head->next == NULL) {
        return head; 
    }
    while(head != NULL && head->next != NULL) {
        prev=head;
        next=head->next;
        temp->val=next->val;
        next->val=prev->val;
        prev->val=temp->val;

        head->next = next->next;
        
    }
    head=curr;
    return head;


}

int main() {
    int arr[] = {1, 2, 3, 4};
    int n = sizeof(arr) / sizeof(arr[0]);
    struct ListNode* head = NULL;
    struct ListNode* temp = NULL;
    struct ListNode* curr = NULL;

    // Dynamically build the linked list from array
    for (int i = 0; i < n; i++) {
        temp = (struct ListNode*)malloc(sizeof(struct ListNode));
        temp->val = arr[i];
        temp->next = NULL;
        if (head == NULL) {
            head = temp;
            curr = head;
        } else {
            curr->next = temp;
            curr = curr->next;
        }
    }

    // Print the linked list
    printList(head);

    // Swap pairs in the linked list
    head = swapPairs(head);
    // Print the modified linked list
    printList(head);

    // Free memory
    curr = head;
    while (curr != NULL) {
        temp = curr;
        curr = curr->next;
        free(temp);
    }

    return 0;
}

File: 24.py
# Node of a Linked List
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


# Linked List class
class LinkedList:
    def __init__(self):
        self.head = None

    # Create linked list from array
    def create_from_list(self, arr):
        if not arr:
            return

        self.head = Node(arr[0])
        current = self.head

        for value in arr[1:]:
            current.next = Node(value)
            current = current.next

    # Print linked list
    def print_list(self):
        current = self.head
        while current:
            print(current.val, end=" -> ")
            current = current.next
        print("None")

    def swap_nodes(self):
        # Dummy node to handle edge cases (like swapping from head)
        dummy = Node(0)
        dummy.next = self.head

        # prev points to the node before the current pair
        prev = dummy

        # Traverse while there are at least two nodes to swap
        while prev.next and prev.next.next:
            # First node of the pair
            first = prev.next

            # Second node of the pair
            second = first.next

            # Perform the swap
            first.next = second.next   # first points to node after the pair
            second.next = first        # second points to first
            prev.next = second         # prev points to second (new head of pair)

            # Move prev to the end of the swapped pair
            prev = first

        # Update head to new starting node
        self.head = dummy.next



head = [1, 2, 3, 4]

ll = LinkedList()
ll.create_from_list(head)
ll.print_list()
ll.swap_nodes()
ll.print_list()

